go-controller/pkg/factory/factory.go

```diff
@@ -52,10 +63,110 @@ func NewWatchFactory(c kubernetes.Interface, stopChan <-chan struct{}) (*WatchFa
+					returnObj, ok = obj.(*kapi.Service)
+				case typeEndpoints:
+					returnObj, ok = obj.(*kapi.Endpoints)
+				case typePolicies:
+					returnObj, ok = obj.(*kapisnetworking.NetworkPolicy)
+				case typeNamespaces:
+					returnObj, ok = obj.(*kapi.Namespace)
+				case typeNodes:
+					returnObj, ok = obj.(*kapi.Node)
+			}
+			if !ok {
+				tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+				if !ok {
+					logrus.Errorf("couldn't get object from tombstone %+v", obj)
+					return
+				}
+				switch informerType {
+					case typePods:
+						returnObj, ok = tombstone.Obj.(*kapi.Pod)
+					case typeServices:
+						returnObj, ok = tombstone.Obj.(*kapi.Service)
+					case typeEndpoints:
+						returnObj, ok = tombstone.Obj.(*kapi.Endpoints)
+					case typePolicies:
+						returnObj, ok = tombstone.Obj.(*kapisnetworking.NetworkPolicy)
+					case typeNamespaces:
+						returnObj, ok = tombstone.Obj.(*kapi.Namespace)
+					case typeNodes:
+						returnObj, ok = tombstone.Obj.(*kapi.Node)
+				}
+				if !ok {
+					logrus.Errorf("tombstone contained object that is not an object of type '%s' : %#v", informerType, obj)
+					return
+				}
+			}
+			wf.FederateDeleteEvent(informerType, returnObj)
+		},
+	}
+}
+
+func (wf *WatchFactory) initializeHandlers() {
+	informerTypes := []string {typePods, typeServices, typeEndpoints, typePolicies, typeNamespaces, typeNodes}
+	for _, informerType := range informerTypes {
+		fh := federatedHandler{
+					mainHandler: wf.defaultHandlerFuncs(informerType),
+					handlerMap: make(map[string]cache.ResourceEventHandler),
+					handlerMutex: &sync.Mutex{},
+				}
+		wf.handlers[informerType] = fh
+		wf.informers[informerType].AddEventHandler(fh.mainHandler)
+	}
+}
+
+func (wf *WatchFactory) FederateAddEvent(informerType string, obj interface {}) {
+	handler := wf.handlers[informerType]
+	handler.handlerMutex.Lock()
+	defer handler.handlerMutex.Unlock()
+	for key, handler := range handler.handlerMap {
+		logrus.Debugf("Issuing Add '%s' event for handler key '%s'", informerType, key)
+		handler.OnAdd(obj)
+	}
+}
+
+func (wf *WatchFactory) FederateUpdateEvent(informerType string, oldObj, newObj interface {}) {
+	handler := wf.handlers[informerType]
+	handler.handlerMutex.Lock()
+	defer handler.handlerMutex.Unlock()
+	for key, handler := range handler.handlerMap {
+		logrus.Debugf("Issuing Update '%s' event for handler key '%s'", informerType, key)
+		handler.OnUpdate(oldObj, newObj)
+	}
+}
+
+func (wf *WatchFactory) FederateDeleteEvent(informerType string, obj interface {}) {
+	handler := wf.handlers[informerType]
+	handler.handlerMutex.Lock()
+	defer handler.handlerMutex.Unlock()
+	for key, handler := range handler.handlerMap {
+		logrus.Debugf("Issuing Delete '%s' event for handler key '%s'", informerType, key)
+		handler.OnDelete(obj)
+	}
+}
+
+func (wf *WatchFactory) addHandler(informerType string, handlerKey string, handler cache.ResourceEventHandler, processExisting func([]interface{})) {
```

Comment:
Also doesn't look like we're really using handlerKey yet, maybe drop that bit until the patch that uses it?  What was your thought there, would it be used as cookie to remove the handler when the callee gofunc is terminated?

Reply:
Yes. That key is for removing the corresponding resource handler when it's done watching. The whole point of this exercise is that only.

Will put it in the relevant commit unless I end up squashing the entire commit.